### PR TITLE
Improve message on view out of bounds access and always abort

### LINKED ISSUE #6851: Consider aborting instead of throwing an exception in view out-of-bound accesses
(Discussion started by Stan on Slack)

Throwing an exception makes things awkward when using multiple MPI ranks.
There does not seem to be a good reason to throw form the host side.
The code that does the out-of-bound check is there
https://github.com/kokkos/kokkos/blob/04a5334c699cb9b87293d27bc73090b3b7c13019/core/src/impl/Kokkos_ViewMapping.hpp#L4016-L4023

We should just call abort instead of throwing.

### PR DESCRIPTION
**Pull request by dalg24**
Fix #6851 

* Always call abort when out-of-bounds access is detected.  (We were throwing a runtime exception when it occurred on the host side.)
* Improve the error message

Before this PR
```
View bounds error of view my_label ( 10 < 1 , 0 < 2 )
```
changed to
```
Kokkos::View ERROR: out of bounds access (label="mylabel") with indices [10,0] but extents [1,2]
```

TLDR; Stan's wishes are my command

### CODE REVIEW

**Comment by dalg24** [comment]
(I am unsure about the code to deduce the view label so I pushed only the tests that I expect to fail to confirm what the current behavior, i.e. before this PR, is)

**Comment by stanmoore1** [comment]
I'm a little concerned with this change: [Do not bother with the label on the device side](https://github.com/kokkos/kokkos/pull/6861/commits/40348e88d265f7dfaf42354db5be18db21d55808). I originally asked for this enhancement in https://github.com/kokkos/kokkos/issues/870. I know it isn't general, but having the label in the error message on the GPU even in some cases is very helpful.

File: core/unit_test/TestViewOutOfBoundsAccess.hpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  using V2 = Kokkos::View<int**,       ExecutionSpace>;
+  using V3 = Kokkos::View<int***,      ExecutionSpace>;
+  using V4 = Kokkos::View<int****,     ExecutionSpace>;
+  using V5 = Kokkos::View<int*****,    ExecutionSpace>;
+  using V6 = Kokkos::View<int******,   ExecutionSpace>;
+  using V7 = Kokkos::View<int*******,  ExecutionSpace>;
+  using V8 = Kokkos::View<int********, ExecutionSpace>;
+  constexpr bool is_device =
+      std::is_same_v<ExecutionSpace, Kokkos::DefaultExecutionSpace> &&
+      std::is_same_v<Kokkos::DefaultExecutionSpace,
```
Comment (dalg24) [inline core/unit_test/TestViewOutOfBoundsAccess.hpp:113]: I had gotten it wrong anyway it should have been and default exec is NOT the same as default host exec

**Comment by dalg24** [comment]
> I'm a little concerned with this change: [Do not bother with the label on the device side](https://github.com/kokkos/kokkos/pull/6861/commits/40348e88d265f7dfaf42354db5be18db21d55808). I originally asked for this enhancement of printing out device side labels in #870. I know it isn't general, but having the label in the error message on the GPU even in some cases is very helpful.

This is fair.  The rational for not bothering with it was that the label info being printed will be garbage if indexing into a subview there the pointer was shifted from the start of the original view.
I reverted the changes.

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 2]
```diff
@@ -3537,110 +3537,90 @@ class ViewMapping<
...
-         view_verify_operator_bounds<R + 1>(map, args...);
+template <class... Indices>
+KOKKOS_FUNCTION constexpr char* append_formated_multidimensional_index(
+    char* dest, Indices... indices) {
+  char* d = dest;
+  Kokkos::Impl::strcat(d, "[");
+  (
+      [&] {
+        d += Kokkos::Impl::strlen(d);
+        (void)Kokkos::Impl::to_chars_i(d,
```
Comment (masterleinad) [inline core/src/impl/Kokkos_ViewMapping.hpp:3555]: Why do you cast to `void` here? Does any compiler complain about the unused return type?

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 3]
```diff
@@ -3537,110 +3537,90 @@ class ViewMapping<
...
+                                       d + 20,  // 20 digits ought to be enough
+                                       indices);
+        Kokkos::Impl::strcat(d, ",");
+      }(),
+      ...);
```
Comment (masterleinad) [inline core/src/impl/Kokkos_ViewMapping.hpp:3560]: Why not just
```suggestion
      for(index: {indices...}) {
        d += Kokkos::Impl::strlen(d);
        Kokkos::Impl::to_chars_i(d,
                                       d + 20,  // 20 digits ought to be enough
                                       index);
        Kokkos::Impl::strcat(d, ",");
      }
```
for a little more "standard" syntax. If we really want to allow different types we can cast `indices` in the initializer list to `std::size_t` or so.

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 4]
```diff
@@ -3537,110 +3537,90 @@ class ViewMapping<
...
-inline void view_error_operator_bounds(char* buf, int len, const MapType& map,
-                                       const iType& i, Args... args) {
-  const int n = snprintf(
-      buf, len, " %ld < %ld %c", static_cast<unsigned long>(i),
-      static_cast<unsigned long>(map.extent(R)), (sizeof...(Args) ? ',' : ')'));
-  view_error_operator_bounds<R + 1>(buf + n, len - n, map, args...);
+template <class Map, class... Indices, std::size_t... Enumerate>
+KOKKOS_FUNCTION void print_extents(char* dest, Map const& map,
+                                   std::index_sequence<Enumerate...>) {
+  (void)append_formated_multidimensional_index(dest, map.extent(Enumerate)...);
```
Comment (masterleinad) [inline core/src/impl/Kokkos_ViewMapping.hpp:3568]: Why `(void)`?

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 5]
```diff
@@ -3537,110 +3537,90 @@ class ViewMapping<
...
+    Kokkos::Impl::strcat(err, "Kokkos::View ERROR: out of bounds access");
+    Kokkos::Impl::strcat(err, " label=(\"");
+    KOKKOS_IF_ON_HOST([&] {
+      if (tracker.m_tracker.has_record()) {
+        Kokkos::Impl::strncat(
+            err, tracker.m_tracker.template get_label<void>().c_str(), 128);
+        return;
+      }
+      Kokkos::Impl::strcat(err, "**UNMANAGED**");
+    }();)
```
Comment (masterleinad) [inline core/src/impl/Kokkos_ViewMapping.hpp:3590]: We don't use immediately invoked lambdas elsewhere with `KOKKOS_IF_ON_HOST`. Did you introduce it because of the extra control flow (`return;`) used below?

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 6]
```diff
@@ -3537,110 +3537,90 @@ class ViewMapping<
 namespace Kokkos {
 namespace Impl {
 
-template <unsigned, class MapType>
-KOKKOS_INLINE_FUNCTION bool view_verify_operator_bounds(const MapType&) {
-  return true;
+template <class Map, class... Indices, std::size_t... Enumerate>
+KOKKOS_FUNCTION bool check_bounds(Map const& map,
```
Comment (masterleinad) [inline core/src/impl/Kokkos_ViewMapping.hpp:3541]:
```suggestion
KOKKOS_FUNCTION bool bounds_satisfied(Map const& map,
```
to make it clearer that we are returning a `bool`.

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 7]
```diff
@@ -3537,110 +3537,90 @@ class ViewMapping<
...
-           (we won't use it, but if its not there then there isn't
-            a corresponding SharedAllocationHeader containing a label).
-           This check should cover the case of Views that don't
-           have the Unmanaged trait but were initialized by pointer. */
-        if (tracker.m_tracker.has_record()) {
-          operator_bounds_error_on_device(map);
-        } else { Kokkos::abort("View bounds error"); }))
+  if (!check_bounds(map, std::make_index_sequence<sizeof...(Args)>(),
+                    args...)) {
+    char err[256] = "";
```
Comment (masterleinad) [inline core/src/impl/Kokkos_ViewMapping.hpp:3580]: What is the impact on kernel size? Do we only get here in `DEBUG` mode? I guess it doesn't matter much since we only get here with `KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK` anyway.

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 8]
```diff
@@ -3537,110 +3537,90 @@ class ViewMapping<
...
+      [&] {
+        d += Kokkos::Impl::strlen(d);
+        (void)Kokkos::Impl::to_chars_i(d,
+                                       d + 20,  // 20 digits ought to be enough
+                                       indices);
+        Kokkos::Impl::strcat(d, ",");
+      }(),
+      ...);
+  d[Kokkos::Impl::strlen(d) - 1] = ']';  // overwrite trailing comma
+  return dest;
```
Comment (masterleinad) [inline core/src/impl/Kokkos_ViewMapping.hpp:3562]: Is the return value used anywhere?

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 9]
```diff
@@ -3537,110 +3537,90 @@ class ViewMapping<
...
+template <class... Indices>
+KOKKOS_FUNCTION constexpr char* append_formated_multidimensional_index(
+    char* dest, Indices... indices) {
+  char* d = dest;
+  Kokkos::Impl::strcat(d, "[");
+  (
+      [&] {
+        d += Kokkos::Impl::strlen(d);
+        (void)Kokkos::Impl::to_chars_i(d,
+                                       d + 20,  // 20 digits ought to be enough
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3556]: Because it can represent the largest representable signed 64-bit integer.
How would you call that named variable?  I thought the comment was good enough but I am open to change if you have a suggestion that improves readability.

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 2 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3555]: Out of habit.  I have not actually tried without it.  Do you want me to drop the cast to `void`?

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 4 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3568]: Out of habit.  I do not know if any compiler actually warns.  Do you want me to remove?

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 5 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3590]: I introduced it to reduce the complexity of the control flow in the device side and I decided to use it as well for the host code where the improvement is less obvious but I judged that consistency with the device code was a good thing.

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 7 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3580]: No impact unless you enable debug_bound_check at configuration time.

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 8 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3562]: Not in the current version.  I chose to return the char pointer argument because that was what a number of string manipulation routine do (e.g. strcat)

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 4 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_ViewMapping.hpp:3568]: Yeah, I would prefer that. AFAIK you would only ever get a warning from ignoring a return value if the function is declared with the [[nodiscard]] attribute.

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3556]: I will change it but I am not a big fan of pulling in `<limits>` into our headers (admittedly we already to that from the exec policy header).
Also I think that it does not necessarily convey that at the time that code was written, we picked a number of digits that we think will suffice whereas I thought the code as I initially wrote does.

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3556]: On 2nd thought, not changing unless more reviewers ask for it.

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 3 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3560]: Also the current implementation would yield
```C++
  Kokkos::View<int**> v("v",2,2);
  std::cout << v(1u, -1);  // Kokkos::View ERROR: out of bounds access label=("v") with indices [1,-1] but extents [2,2]
```
which is better than `Kokkos::View ERROR: out of bounds access label=("v") with indices [1,18446744073709551615] but extents [2,2]`

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 6 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp]: I renamed it to `within_range`.  Let me know if you still do not like the name.

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 8 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3562]: I was not planning to change.  Please unresolved and comment if you want me to change.

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 3 above)
Comment (crtrott) [reply core/src/impl/Kokkos_ViewMapping.hpp]: agreed

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 10]
```diff
@@ -3537,110 +3537,88 @@ class ViewMapping<
...
 namespace Impl {
 
-template <unsigned, class MapType>
-KOKKOS_INLINE_FUNCTION bool view_verify_operator_bounds(const MapType&) {
-  return true;
+template <class Map, class... Indices, std::size_t... Enumerate>
+KOKKOS_FUNCTION bool within_range(Map const& map,
+                                  std::index_sequence<Enumerate...>,
+                                  Indices... indices) {
+  return (... && ((std::size_t)indices < map.extent(Enumerate)));
```
Comment (crtrott) [inline core/src/impl/Kokkos_ViewMapping.hpp:3544]: still stumbling slightly over the left fold, but no reason to block

**Review by crtrott — APPROVED** [review]
Looks largely fine to me. I kinda agree with Daniel on the host lambda thing, don't see a point in it, and I am also stumbling over the left fold. But its not like I care that much

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 10 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp:3544]: Oh yes sorry I had agreed to change

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 5 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp]: here is what it looks like without the lambda trick
```diff
@@ -3580,13 +3580,12 @@ KOKKOS_INLINE_FUNCTION void view_verify_operator_bounds(
     char err[256] = "";
     strcat(err, "Kokkos::View ERROR: out of bounds access");
     strcat(err, " label=(\"");
-    KOKKOS_IF_ON_HOST([&] {
-      if (tracker.m_tracker.has_record()) {
-        strncat(err, tracker.m_tracker.template get_label<void>().c_str(), 128);
-        return;
-      }
-      strcat(err, "**UNMANAGED**");
-    }();)
+    KOKKOS_IF_ON_HOST(
+        if (tracker.m_tracker.has_record()) {
+          strncat(err, tracker.m_tracker.template get_label<void>().c_str(),
+                  128);
+          return;
+        } else { strcat(err, "**UNMANAGED**"); })
     KOKKOS_IF_ON_DEVICE([&] {
       // Check #1: is there a SharedAllocationRecord?  (we won't use it, but
       // if its not there then there isn't a corresponding

Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp]: Anyway I changed it since both Christian and you complained about it

File: core/src/impl/Kokkos_ViewMapping.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/impl/Kokkos_ViewMapping.hpp]:
```diff
diff --git a/core/src/impl/Kokkos_ViewMapping.hpp b/core/src/impl/Kokkos_ViewMapping.hpp
index 89eff343c..b4894a8ec 100644
--- a/core/src/impl/Kokkos_ViewMapping.hpp
+++ b/core/src/impl/Kokkos_ViewMapping.hpp
@@ -3549,15 +3549,12 @@ KOKKOS_FUNCTION constexpr char* append_formatted_multidimensional_index(
     char* dest, Indices... indices) {
   char* d = dest;
   strcat(d, "[");
-  (
-      [&] {
-        d += strlen(d);
-        to_chars_i(d,
-                   d + 20,  // 20 digits ought to be enough
-                   (size_t)indices);
-        strcat(d, ",");
-      }(),
-      ...);
+  for (auto index : {static_cast<size_t>(indices)...}) {
+    d += strlen(d);
+    constexpr int max_digits = 20;  // enough for uint64_t
+    to_chars_i(d, d + max_digits, index);
+    strcat(d, ",");
+  }
   d[strlen(d) - 1] = ']';  // overwrite trailing comma
   return dest;
 }
```
would also look better to me.

[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/impl/Kokkos_ViewMapping.hpp]: or just the `max_digits` part.

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/impl/Kokkos_ViewMapping.hpp]: The `static_cast` would potentially lead to poor quality error messages (see https://github.com/kokkos/kokkos/pull/6861#discussion_r1517992352).
I don't think the `max_digits` is better than the current version and I am not planning to change unless more reviewers ask me to change it.

**Review by Rombur — APPROVED** [review]

**Comment by dalg24** [comment]
Unrelated runtime failures in the ROCm 5.2 HIP build

**Comment by dalg24** [comment]
Ignoring
```
[100%] Linking Fortran executable example
/usr/bin/ld: /usr/bin/../lib/gcc/x86_64-redhat-linux/14/../../../../lib64/libFortran_main.a(Fortran_main.c.o): in function `main':
(.text.startup.main+0x0): multiple definition of `main'; CMakeFiles/example.dir/cmake_example.cpp.o:cmake_example.cpp:(.text+0x0): first defined here
/usr/bin/ld: /usr/bin/../lib/gcc/x86_64-redhat-linux/14/../../../../lib64/libFortran_main.a(Fortran_main.c.o): in function `main':
(.text.startup.main+0xb): undefined reference to `_QQEnvironmentDefaults'
/usr/bin/ld: (.text.startup.main+0x18): undefined reference to `_QQmain'
flang-new: error: linker command failed with exit code 1 (use -v to see invocation)
gmake[2]: *** [CMakeFiles/example.dir/build.make:119: example] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/example.dir/all] Error 2
gmake: *** [Makefile:101: all] Error 2
Error: Process completed with exit code 2.
``` in Fedora rawhide builds
